ddl/check.lua

```diff
@@ -628,75 +628,147 @@ local function check_index(i, index, space)
+            "spaces[%q].format[%q].type: bad value (unsigned expected, got %s)",
+            space.name, 'bucket_id', space.fields.bucket_id.type
+        )
+    end
+
+    return true
+end
+
+local function check_bucket_id_index(space)
+    if not space.indexes.bucket_id then
+        return nil, string.format(
+            "spaces[%q].indexes: sharding_key exists in the space, but there's" ..
+            " no bucket_id defined in 'indexes' section",
+            space.name
+        )
+    end
+
+    if space.indexes.bucket_id.unique then
+        return nil, string.format(
+            "spaces[%q].indexes[%q].unique: bucket_id index can't be unique",
+            space.name, 'bucket_id'
+        )
+    end
+
+    if #space.indexes['bucket_id'].parts ~= 1 then
+        return nil, string.format(
+            "spaces[%q].indexes[%q].parts: bucket_id index can't be composite (1 part expected, got %d parts)",
+            space.name, 'bucket_id', #space.indexes['bucket_id'].parts
+        )
+    end
+
+    if space.indexes['bucket_id'].parts[1].path ~= 'bucket_id' then
+        return nil, string.format(
+            "spaces[%q].indexes[%q].parts[1].path: invalid field reference " ..
+            "(reference to bucket_id expected, got %s)",
+            space.name, 'bucket_id', space.indexes['bucket_id'].parts[1].path
+        )
+    end
+
+    return true
+end
+
+local function is_callable(object)
+    if type(object) == 'function' then
         return true
     end
 
-    do -- check that format['bucket_id'] valid
-        if not space.fields.bucket_id then
-            return nil, string.format(
-                "spaces[%q].format: sharding_key exists in the space, but there's" ..
-                " no bucket_id defined in 'format' section",
-                space.name
-            )
-        end
+    -- all objects with type `cdata` are allowed
+    -- because there is no easy way to get
+    -- metatable.__call of object with type `cdata`
+    if type(object) == 'cdata' then
+        return true
+    end
 
-        if space.fields.bucket_id.type ~= 'unsigned' then
-            return nil, string.format(
-                "spaces[%q].format[%q].type: bad value (unsigned expected, got %s)",
-                space.name, 'bucket_id', space.fields.bucket_id.type
-            )
-        end
+    local object_metatable = getmetatable(object)
+    if (type(object) == 'table' or type(object) == 'userdata')
+    and object_metatable and object_metatable.__call then
+        -- `__call` metamethod can be only the `function`
+        -- and cannot be a `table` | `userdata` | `cdata`
+        -- with `__call` methamethod on its own
+        return type(object_metatable.__call) == 'function'
+    end
+
+    return false
+end
+
+local function check_custom_sharding_func(sharding_func_name)
+    local sharding_func = rawget(_G, sharding_func_name)
+    if sharding_func == nil then
+        return false
+    end
+
+    return is_callable(sharding_func)
+end
```

Comment:
Why we use different rules for single component names and multiple dot separated component ones?

1. We accept any string for a single component name (say, starting from a digit). But not for components of multicomponent names.
2. We check whether a single component name is callable, but we don't do that for multicomponent one.

I don't thnk that there is any sense to divide names to 'custom' (single component) and 'module' (multiple component). A user can easily register a multicomponent name:

```lua
_G.foo = {}
function foo.bar(<...>)
    <...>
end
```

The single component name is the case of multicomponent one (with one component).

In fact, we know nothing about modules like vshard. We just perform `_G.<component 1>.<component 2>.<..and so on..>` to obtain a function.